File: next/dweb-browser/src/Platforms/iOS/MicroModule/Browser/Jmm/JmmNMM.cs
using System.Net;
using System.Web;
using DwebBrowser.MicroService.Http;

namespace DwebBrowser.MicroService.Browser.Jmm;

public class JmmNMM : NativeMicroModule
{
    static readonly Debugger Console = new("JmmNMM");

    static JmmNMM()
    {
        NativeFetch.NativeFetchAdaptersManager.Append(GetUsrFile);
    }

    public static async Task<PureResponse?> GetUsrFile(MicroModule remote, PureRequest request)
    {

        if (request.ParsedUrl is not null and var parsedUrl &&
            parsedUrl.Scheme == Uri.UriSchemeFile &&
            parsedUrl.FullHost is "" &&
            parsedUrl.Path.StartsWith("/usr/"))
        {
            var query = HttpUtility.ParseQueryString(parsedUrl.Query);
            var mode = query["mode"] ?? "auto";
            //var chunk = query["chunk"]?.ToIntOrNull() ?? 1024 * 1024;

            var relativePath = parsedUrl.Path;

            var baseDir = JsMicroModule.GetInstallPath(remote.Mmid, remote.Version);
            return await LocaleFile.ReadLocalFileAsResponse(baseDir, relativePath, mode, url: request.Url);
        }

        return null;

    }

    public override List<Dweb_DeepLink> Dweb_deeplinks { get; init; } = new() { "dweb:install" };
    public override List<MicroModuleCategory> Categories { get; init; } = new()
    {
        MicroModuleCategory.Service,
        MicroModuleCategory.Hub_Service,
    };

    public static JmmController JmmController = null!;

    public override string ShortName { get; set; } = "JMM";
    public JmmNMM() : base("jmm.browser.dweb", "Js MicroModule Management")
    {
    }

    protected override async Task _bootstrapAsync(IBootstrapContext bootstrapContext)
    {
        InstallJmmApps();

        HttpRouter.AddRoute(IpcMethod.Get, "/install", async (request, ipc) =>
        {
            var searchParams = request.SafeUrl.SearchParams;
            var metadataUrl = request.QueryStringRequired("url");
            var jmmMetadata = await (await NativeFetchAsync(metadataUrl)).JsonAsync<JmmAppInstallManifest>();
            var url = new URL(metadataUrl);

            if (jmmMetadata is JmmAppInstallManifest metadata)
            {
                await OpenJmmMetadataInstallPage(metadata, ipc, url);
            }

            return jmmMetadata;
        });

        HttpRouter.AddRoute(IpcMethod.Get, "/uninstall", async (request, _) =>
        {
            var searchParams = request.SafeUrl.SearchParams;
            var mmid = searchParams.ForceGet("app_id");
            await OpenJmmMetadataUninstallPage(mmid);

            return new PureResponse(HttpStatusCode.OK, Body: new PureUtf8StringBody(@"{""ok"":true}"));
        });

        HttpRouter.AddRoute(IpcMethod.Get, "/closeApp", async (request, _) =>
        {
            var mmid = request.QueryStringRequired("app_id");
            await JmmController.CloseApp(mmid);
            return true;
        });

        // App详情
        HttpRouter.AddRoute(IpcMethod.Get, "/detailApp", async (request, ipc) =>
        {
            var mmid = request.QueryStringRequired("app_id");
            var microModule = await bootstrapContext.Dns.Query(mmid);

            if (microModule is null)
            {
                return new PureResponse(HttpStatusCode.NotFound, Body: new PureUtf8StringBody("not found " + mmid));
            }

            if (microModule is JsMicroModule jsMicroModule)
            {
                var metadata = jsMicroModule.Metadata.Config;
                var jmmAppDownloadManifest = JmmAppDownloadManifest.FromInstallManifest(metadata);
                jmmAppDownloadManifest.DownloadStatus = GetCurrentDownloadStatus(metadata);

                await MainThread.InvokeOnMainThreadAsync(async () => await OpenJmmMetadataInstallPage(jmmAppDownloadManifest, ipc));

                return new PureResponse(HttpStatusCode.OK, Body: new PureUtf8StringBody(@"{""ok"":true}"));
            }

            return new PureResponse(HttpStatusCode.NotFound, Body: new PureUtf8StringBody("not found " + mmid));
        });

        HttpRouter.AddRoute(IpcMethod.Get, "/pause", async (_, ipc) =>
        {
            return JmmDwebService.UpdateDownloadControlStatus(ipc.Remote.Mmid, DownloadControlStatus.Pause);
        });

        HttpRouter.AddRoute(IpcMethod.Get, "/resume", async (_, ipc) =>
        {
            return JmmDwebService.UpdateDownloadControlStatus(ipc.Remote.Mmid, DownloadControlStatus.Resume);
        });

        HttpRouter.AddRoute(IpcMethod.Get, "/cancel", async (_, ipc) =>
        {
            return JmmDwebService.UpdateDownloadControlStatus(ipc.Remote.Mmid, DownloadControlStatus.Cancel);
        });
    }

    private DownloadStatus GetCurrentDownloadStatus(IJmmAppInstallManifest manifest)
    {
        var oldAmmMetadata = JmmDatabase.Instance.Find(manifest.Id);
        var initDownloadStatus = DownloadStatus.IDLE;

        if (oldAmmMetadata is not null)
        {
            var oldSemver = new Semver(oldAmmMetadata.Version);
            var newSemver = new Semver(manifest.Version);

            if (newSemver.CompareTo(oldSemver) > 0)
            {
                initDownloadStatus = DownloadStatus.NewVersion;
            }
            else if (newSemver.CompareTo(oldSemver) == 0)
            {
                initDownloadStatus = DownloadStatus.Installed;
            }
        }

        return initDownloadStatus;
    }

    private async Task OpenJmmMetadataInstallPage(JmmAppInstallManifest manifest, Ipc ipc, URL? url = null)
    {
        if (!manifest.BundleUrl.StartsWith(Uri.UriSchemeHttp) && !manifest.BundleUrl.StartsWith(Uri.UriSchemeHttps))
        {
            if (url is not null)
            {
                manifest.BundleUrl = (new URL(new Uri(url.Uri, manifest.BundleUrl))).Href;
            }
        }

        try
        {
            await MainThread.InvokeOnMainThreadAsync(async () =>
            {
                var win = await WindowAdapterManager.Instance.CreateWindow(
                    new WindowState(ipc.Remote.Mmid, Mmid) { Mode = WindowMode.MAXIMIZE });

                JmmController = new JmmController(this, win);

                var jmmAppDownloadManifest = JmmAppDownloadManifest.FromInstallManifest(manifest);
                jmmAppDownloadManifest.DownloadStatus = GetCurrentDownloadStatus(manifest);
                await JmmController.OpenDownloadPageAsync(jmmAppDownloadManifest);
            });
        }
        catch (Exception e)
        {
            Console.Log("_openJmmMetadataInstallPage", e.Message);
            Console.Log("_openJmmMetadataInstallPage", e.StackTrace);
        }
    }

    private async Task OpenJmmMetadataUninstallPage(Mmid mmid)
    {
        BootstrapContext.Dns.UnInstall(mmid);
        JmmDwebService.UnInstall(mmid);
        JmmDatabase.Instance.Remove(mmid);
    }

    /// <summary>
    /// 注册所有已经下载的应用
    /// </summary>
    private void InstallJmmApps()
    {
        _ = Task.Run(() =>
        {
            foreach (var appInfo in JmmDatabase.Instance.All())
            {
                var metadata = new JsMMMetadata(appInfo);
                var jmm = new JsMicroModule(metadata);
                BootstrapContext.Dns.Install(jmm);
            }
        }).NoThrow();
    }
}

File: android/app/src/main/java/info/bagen/dwebbrowser/microService/browser/jmm/JmmNMM.kt
package info.bagen.dwebbrowser.microService.browser.jmm

import info.bagen.dwebbrowser.App
import io.ktor.http.HttpMethod
import io.ktor.http.HttpStatusCode
import io.ktor.http.URLBuilder
import io.ktor.http.Url
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import org.dweb_browser.browserUI.database.AppType
import org.dweb_browser.browserUI.database.DeskAppInfo
import org.dweb_browser.browserUI.database.DeskAppInfoStore
import org.dweb_browser.browserUI.download.DownLoadController
import org.dweb_browser.browserUI.download.isGreaterThan
import org.dweb_browser.browserUI.microService.browser.link.WebLinkMicroModule
import org.dweb_browser.browserUI.util.APP_DIR_TYPE
import org.dweb_browser.browserUI.util.BrowserUIApp
import org.dweb_browser.browserUI.util.FilesUtil
import org.dweb_browser.helper.Debugger
import org.dweb_browser.helper.ImageResource
import org.dweb_browser.helper.resolvePath
import org.dweb_browser.microservice.core.AndroidNativeMicroModule
import org.dweb_browser.microservice.core.BootstrapContext
import org.dweb_browser.microservice.help.types.IMicroModuleManifest
import org.dweb_browser.microservice.help.types.JmmAppInstallManifest
import org.dweb_browser.microservice.help.types.MICRO_MODULE_CATEGORY
import org.dweb_browser.microservice.help.types.MMID
import org.dweb_browser.microservice.http.PureResponse
import org.dweb_browser.microservice.http.bind
import org.dweb_browser.microservice.http.bindDwebDeeplink
import org.dweb_browser.microservice.ipc.Ipc
import org.dweb_browser.microservice.sys.dns.RespondLocalFileContext.Companion.respondLocalFile
import org.dweb_browser.microservice.sys.dns.nativeFetch
import org.dweb_browser.microservice.sys.dns.nativeFetchAdaptersManager
import org.dweb_browser.microservice.sys.dns.returnAndroidFile
import org.dweb_browser.window.core.WindowState
import org.dweb_browser.window.core.constant.WindowConstants
import org.dweb_browser.window.core.constant.WindowMode
import org.dweb_browser.window.core.createWindowAdapterManager
import org.dweb_browser.window.core.helper.setFromManifest
import java.io.File

val debugJMM = Debugger("JMM")

/**
 * 获取 map 值，如果不存在，则使用defaultValue; 如果replace 为true也替换为defaultValue
 */
inline fun <K, V> MutableMap<K, V>.getOrPutOrReplace(
  key: K, replaceValue: (V) -> V, defaultValue: () -> V
): V {
  val value = get(key)
  return if (value == null) {
    val answer = defaultValue()
    put(key, answer)
    answer
  } else {
    replaceValue(value)
  }
}

class JmmNMM : AndroidNativeMicroModule("jmm.browser.dweb", "Js MicroModule Management") {
  init {
    short_name = "JMM";
    dweb_deeplinks = listOf("dweb://install")
    categories = listOf(MICRO_MODULE_CATEGORY.Service, MICRO_MODULE_CATEGORY.Hub_Service);
    icons = listOf(
      ImageResource(
        src = "file:///sys/icons/$mmid.svg", type = "image/svg+xml", purpose = "monochrome"
      )
    )

    /// 提供JsMicroModule的文件适配器
    /// 这个适配器不需要跟着bootstrap声明周期，只要存在JmmNMM模块，就能生效
    nativeFetchAdaptersManager.append { fromMM, request ->
      return@append request.respondLocalFile {
        if (filePath.startsWith("/usr/")) {
          debugJMM("UsrFile", "$fromMM => ${request.href}")
          returnAndroidFile(
            getAppContext().dataDir.absolutePath + File.separator + APP_DIR_TYPE.SystemApp.rootName + File.separator + fromMM.mmid,
            filePath
          )
        } else returnNext()
      }
    }
  }

  enum class EIpcEvent(val event: String) {
    State("state"), Ready("ready"), Activity("activity"), Close("close")
  }

  private var jmmController: JmmController? = null

  fun getApps(mmid: MMID): IMicroModuleManifest? {
    return bootstrapContext.dns.query(mmid)
  }

  override suspend fun _bootstrap(bootstrapContext: BootstrapContext) {
    installJmmApps()

    this.onAfterShutdown {
      jmmController = null
    }

    val routeInstallHandler = definePureResponse {
      val metadataUrl = request.queryOrFail("url")
      val response = nativeFetch(metadataUrl)
      if (response.isOk()) {
        try {
          val jmmAppInstallManifest = response.json<JmmAppInstallManifest>()
          val url = Url(metadataUrl)
          // 根据 jmmMetadata 打开一个应用信息的界面，用户阅读界面信息后，可以点击"安装"
          installJsMicroModule(jmmAppInstallManifest, ipc, url)
          PureResponse(HttpStatusCode.OK)
        } catch (e: Throwable) {
          PureResponse(HttpStatusCode.ExpectationFailed).body(e.stackTraceToString())
        }
      } else {
        PureResponse(HttpStatusCode.ExpectationFailed).body("invalid status code: ${response.status}")
      }
    }
    routes(
      // 安装
      "install" bindDwebDeeplink routeInstallHandler,
      "/install" bind HttpMethod.Get to routeInstallHandler,
      "/uninstall" bind HttpMethod.Get to defineBooleanResponse {
        val mmid = request.queryOrFail("app_id")
        debugJMM("uninstall", mmid)
        jmmMetadataUninstall(mmid)
        true
      },
      //检查是否有新版本
      "/check" bind HttpMethod.Get to defineBooleanResponse {
        val mmid = request.queryOrFail("app_id")
        true
      },
      // app详情
      "/detailApp" bind HttpMethod.Get to defineBooleanResponse {
        val mmid = request.queryOrFail("app_id")
        debugJMM("detailApp", mmid)
        val microModule = bootstrapContext.dns.query(mmid)

        // TODO: 系统原生应用如WebBrowser的详情页展示？
        if (microModule is JsMicroModule) {
          installJsMicroModule(microModule.metadata, ipc)
          true
        } else {
          false
        }
      },
      "/pause" bind HttpMethod.Get to defineBooleanResponse {
        BrowserUIApp.Instance.mBinderService?.invokeUpdateDownloadStatus(
          ipc.remote.mmid, DownLoadController.PAUSE
        )
        true
      },
      /**继续下载*/
      "/resume" bind HttpMethod.Get to defineBooleanResponse {
        debugJMM("resume", ipc.remote.mmid)
        BrowserUIApp.Instance.mBinderService?.invokeUpdateDownloadStatus(
          ipc.remote.mmid, DownLoadController.RESUME
        )
        true
      },
      "/cancel" bind HttpMethod.Get to defineBooleanResponse {
        debugJMM("cancel", ipc.remote.mmid)
        BrowserUIApp.Instance.mBinderService?.invokeUpdateDownloadStatus(
          ipc.remote.mmid, DownLoadController.CANCEL
        )
        true
      })
  }

  override suspend fun _shutdown() {
  }

  /**
   * 从内存中加载数据
   */
  private fun installJmmApps() {
    ioAsyncScope.launch {
      var preList = mutableListOf<DeskAppInfo>()
      DeskAppInfoStore.queryDeskAppInfoList().collectLatest { list -> // TODO 只要datastore更新，这边就会实时更新
        debugJMM("AppInfoDataStore", "size=${list.size}")
        list.map { deskAppInfo ->
          when (deskAppInfo.appType) {
            AppType.MetaData -> deskAppInfo.metadata?.let { jsMetaData ->
              preList.removeIf { it.metadata?.id == jsMetaData.id }
              // 检测版本
              bootstrapContext.dns.query(jsMetaData.id)?.let { lastMetaData ->
                if (jsMetaData.version.isGreaterThan(lastMetaData.version)) {
                  bootstrapContext.dns.close(lastMetaData.mmid)
                }
              }
              bootstrapContext.dns.install(JsMicroModule(jsMetaData))
            }

            AppType.URL -> deskAppInfo.weblink?.let { deskWebLink ->
              preList.removeIf { preDeskAppInfo -> preDeskAppInfo.weblink?.id == deskWebLink.id }
              bootstrapContext.dns.install(WebLinkMicroModule(deskWebLink))
            }
          }
        }
        /// 将剩余的应用卸载掉
        for (uninstallItem in preList) {
          uninstallItem.weblink?.deleteIconFile() // 删除已下载的图标
          (uninstallItem.metadata?.id ?: uninstallItem.weblink?.id)?.let { uninstallId ->
            bootstrapContext.dns.uninstall(uninstallId)
          }
        }
        preList = list
      }
    }
  }

  private suspend fun installJsMicroModule(
    jmmAppInstallManifest: JmmAppInstallManifest, ipc: Ipc, url: Url? = null,
  ) {
    jmmController?.closeSelf() // 如果存在的话，关闭先，同时可以考虑置空
    if (!jmmAppInstallManifest.bundle_url.startsWith("http")) {
      url?.let {
        jmmAppInstallManifest.bundle_url =
          URLBuilder(it).run {
            resolvePath(jmmAppInstallManifest.bundle_url);
            buildString()
          }
      }
    }
    debugJMM("openJmmMetadataInstallPage", jmmAppInstallManifest.bundle_url)
    // 打开安装的界面
    // JmmManagerActivity.startActivity(jmmAppInstallManifest)
    // 打开安装窗口
    val win = createWindowAdapterManager.createWindow(WindowState(
      WindowConstants(
        owner = mmid, ownerVersion = version, provider = mmid, microModule = this
      )
    ).apply {
      mode = WindowMode.FLOATING
      setFromManifest(this@JmmNMM)
    })
    jmmController = JmmController(win, this@JmmNMM, jmmAppInstallManifest).also { ctrl ->
      ctrl.onClosed {
        if (jmmController == ctrl) {
          jmmController = null
        }
      }
    }
  }

  private suspend fun jmmMetadataUninstall(mmid: MMID) {
    // 先从列表移除，然后删除文件
    bootstrapContext.dns.uninstall(mmid)
    DeskAppInfoStore.deleteDeskAppInfo(mmid)
    FilesUtil.uninstallApp(App.appContext, mmid)
  }
}
